File: backend/app/core/password_validation.py
import re


PASSWORD_REQUIREMENTS_MESSAGE = (
    "Password must be at least 8 characters and include uppercase, "
    "lowercase, number, and symbol."
)


def validate_password_strength(value: str) -> str:
    if len(value) < 8:
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    if not re.search(r"[A-Z]", value):
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    if not re.search(r"[a-z]", value):
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    if not re.search(r"[0-9]", value):
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    if not re.search(r"[^A-Za-z0-9]", value):
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    return value

File: frontend/src/app/lib/passwordValidation.ts
export const PASSWORD_REQUIREMENTS_ERROR =
  'Password must be at least 8 characters and include uppercase, lowercase, number, and symbol.';

export const PASSWORD_REQUIREMENTS = [
  { text: 'At least 8 characters', test: (value: string) => value.length >= 8 },
  { text: 'Contains uppercase letter', test: (value: string) => /[A-Z]/.test(value) },
  { text: 'Contains lowercase letter', test: (value: string) => /[a-z]/.test(value) },
  { text: 'Contains number', test: (value: string) => /[0-9]/.test(value) },
  { text: 'Contains symbol', test: (value: string) => /[^A-Za-z0-9]/.test(value) },
];

export function getPasswordError(password: string) {
  return PASSWORD_REQUIREMENTS.every((requirement) => requirement.test(password))
    ? undefined
    : PASSWORD_REQUIREMENTS_ERROR;
}

export function isStrongPassword(password: string) {
  return getPasswordError(password) === undefined;
}
